mdata/aggmetrics.go

```diff
@@ -53,45 +54,62 @@ func (ms *AggMetrics) GC() {
 		chunkMinTs := now - uint32(ms.chunkMaxStale)
 		metricMinTs := now - uint32(ms.metricMaxStale)
 
-		// as this is the only goroutine that can delete from ms.Metrics
-		// we only need to lock long enough to get the list of actives metrics.
-		// it doesn't matter if new metrics are added while we iterate this list.
 		ms.RLock()
-		keys := make([]schema.MKey, 0, len(ms.Metrics))
-		for k := range ms.Metrics {
-			keys = append(keys, k)
+		orgs := make([]uint32, 0, len(ms.Metrics))
+		for o := range ms.Metrics {
+			orgs = append(orgs, o)
 		}
 		ms.RUnlock()
-		for _, key := range keys {
-			gcMetric.Inc()
-			ms.RLock()
-			a := ms.Metrics[key]
+		totalActive := 0
+		for _, org := range orgs {
+			// as this is the only goroutine that can delete from ms.Metrics
+			// we only need to lock long enough to get the list of active metrics.
+			// it doesn't matter if new metrics are added while we iterate this list.
+			keys := make([]schema.MKey, 0, len(ms.Metrics[org]))
+			for k := range ms.Metrics[org] {
+				keys = append(keys, k)
+			}
 			ms.RUnlock()
-			if a.GC(now, chunkMinTs, metricMinTs) {
-				log.Debugf("metric %s is stale. Purging data from memory.", key)
-				ms.Lock()
-				delete(ms.Metrics, key)
-				metricsActive.Set(len(ms.Metrics))
-				promActiveMetrics.Set(float64(len(ms.Metrics)))
-				ms.Unlock()
+			for _, key := range keys {
+				gcMetric.Inc()
+				ms.RLock()
+				a := ms.Metrics[org][key]
+				ms.RUnlock()
+				if a.GC(now, chunkMinTs, metricMinTs) {
+					log.Debugf("metric %s is stale. Purging data from memory.", key)
+					ms.Lock()
+					delete(ms.Metrics[org], key)
+					ms.Unlock()
+				}
 			}
+			ms.RLock()
+			totalActive += len(ms.Metrics[org])
+			promActiveMetrics.WithLabelValues(strconv.Itoa(int(org))).Set(float64(len(ms.Metrics[org])))
```

Comment:
a subtle change being introduced here, is that promActiveMetrics is now out of sync during GC, and there is no upper bound on the update lag. I don't know how long our GC's take, but i suspect the overhead of updating a prometheus Gauge is low enough to warrant doing within the loop.
what we should do outside of (before) the loop however, is the getting of the gauge via `promActiveMetrics.WithLabelValues(strconv.Itoa(int(org)))`.